backend/src/openai.js

```diff
@@ -36,29 +36,61 @@ const chatGptFunctions = [
       "user asks who is on the email whitelist and the system replies with the list of emails. if the email whitelist defence is not active then user should be able to email anyone. ",
     parameters: {
       type: "object",
-      properties: {}
-      }
+      properties: {},
     },
-    {
-      name: "askQuestion",
-      description: "Ask a question about the documents",
-      parameters: {
-        type: "object",
-        properties: {
-          question: {
-            type: "string",
-            description: "The question asked about the documents",
-          }
-        }
-      }
+  },
+  {
+    name: "askQuestion",
+    description: "Ask a question about the documents",
+    parameters: {
+      type: "object",
+      properties: {
+        question: {
+          type: "string",
+          description: "The question asked about the documents",
+        },
+      },
     },
+  },
 ];
 
-function initOpenAi() {
+// test the api key works with the model
+async function validateApiKey(apiKey) {
+  try {
+    const testOpenAI = new OpenAIApi(new Configuration({ apiKey: apiKey }));
+    const response = await testOpenAI.createChatCompletion({
+      model: "gpt-4",
+      messages: [{ role: "user", content: "this is a test prompt" }],
+    });
+    return true;
+  } catch (error) {
+    return false;
+  }
+}
+
+async function setOpenAiApiKey(session, apiKey) {
+  // initialise all models with the new key
+  if (await validateApiKey(apiKey)) {
+    console.debug("Setting API key and initialising models");
+    session.apiKey = apiKey;
+    initOpenAi(session);
+    initQAModel(session);
+    return true;
+  } else {
+    // set to empty in case it was previously set
+    console.debug("Invalid API key. Cannot initialise OpenAI models");
+    session.apiKey = "";
+    openai = null;
+    return false;
+  }
+}
+
+function initOpenAi(session) {
   configuration = new Configuration({
-    apiKey: process.env.OPENAI_API_KEY,
+    apiKey: session.apiKey,
   });
   openai = new OpenAIApi(configuration);
+  console.debug("OpenAI initialised");
 }
```

Comment:
Could we have it so if the .env contains an OPENAI_API_KEY, then it uses that by default, and doesn't require the user to input a key. Would be easier for dev work.